sycl/include/sycl/ext/intel/experimental/esimd/detail/types.hpp

```diff
@@ -209,85 +136,96 @@ template <typename Ty, int N> struct vector_type {
-struct compute_format_type_2d_impl {
-  static constexpr int Prod = sizeof(Ty) * N / sizeof(EltTy);
-  static_assert(Prod == Width * Height, "size mismatch");
+// @{
+// Converts element type of given simd type \c SimdT to
+// given scalar type \c NewElemT.
+template <class SimdT, typename NewElemT> struct convert_simd_elem_type;
 
-  static constexpr int SizeX = Width;
-  static constexpr int StrideX = 1;
-  static constexpr int SizeY = Height;
-  static constexpr int StrideY = 1;
-  using type = region2d_t<EltTy, SizeY, StrideY, SizeX, StrideX>;
+// Specialization for the simd_obj_impl type.
+template <typename OldElemT, int N, typename NewElemT,
+          template <typename, int> class SimdT>
+struct convert_simd_elem_type<
+    simd_obj_impl<__raw_t<OldElemT>, N, SimdT<OldElemT, N>>, NewElemT> {
+  using type = simd_obj_impl<__raw_t<NewElemT>, N, SimdT<NewElemT, N>>;
 };
 
-template <typename Ty, int N, typename EltTy, int Height, int Width,
+// Specialization for the simd_obj_impl type derivatives.
+template <typename OldElemT, int N, typename NewElemT,
           template <typename, int> class SimdT>
-struct compute_format_type_2d<SimdT<Ty, N>, EltTy, Height, Width>
-    : compute_format_type_2d_impl<Ty, N, EltTy, Height, Width> {};
-
-template <typename Ty, int N, typename EltTy, int Height, int Width,
-          class SimdT>
-struct compute_format_type_2d<simd_obj_impl<Ty, N, SimdT>, EltTy, Height, Width>
-    : compute_format_type_2d_impl<Ty, N, EltTy, Height, Width> {};
-
-template <typename BaseTy, typename RegionTy, typename EltTy, int Height,
-          int Width>
-struct compute_format_type_2d<simd_view<BaseTy, RegionTy>, EltTy, Height,
-                              Width> {
-  using ShapeTy = typename shape_type<RegionTy>::type;
-  static constexpr int Prod = ShapeTy::Size_in_bytes / sizeof(EltTy);
-  static_assert(Prod == Width * Height, "size mismatch");
-
-  static constexpr int SizeX = Width;
-  static constexpr int StrideX = 1;
-  static constexpr int SizeY = Height;
-  static constexpr int StrideY = 1;
-  using type = region2d_t<EltTy, SizeY, StrideY, SizeX, StrideX>;
+struct convert_simd_elem_type<SimdT<OldElemT, N>, NewElemT> {
+  using type = SimdT<NewElemT, N>;
 };
 
-template <typename Ty, typename EltTy, int Height, int Width>
-using compute_format_type_2d_t =
-    typename compute_format_type_2d<Ty, EltTy, Height, Width>::type;
+// Convenience shortcut.
+template <class SimdT, typename NewElemT>
+using convert_simd_elem_type_t =
+    typename convert_simd_elem_type<SimdT, NewElemT>::type;
+
+// @}
+
+// Constructs a simd type with the same template type as in \c SimdT, and
+// given element type and number.
+template <class SimdT, typename T, int N>
+using construct_a_simd_type_t =
+    convert_simd_elem_type_t<resize_a_simd_type_t<SimdT, N>, T>;
+
+// @}
+
+// must match simd_mask<N>::element_type
+template <int N>
+using simd_mask_storage_t = vector_type_t<simd_mask_elem_type, N>;
```

Comment:
No functional changes in the code above (moved here from different part and renamed template parameter)